jb/itch5/array_based_order_book.hpp

```diff
@@ -0,0 +1,617 @@
+      }
+      return false;
+    }
+
+    // handles the top_levels_ price
+    if (better_(px, px_inside_)) {
+      std::ostringstream os;
+      os << "array_based_order_book::reduce_order."
+         << " Trying to reduce a non-existing top_levels_ price"
+         << " (better px_inside)."
+         << " px_begin_top_=" << px_begin_top_ << " px_inside_=" << px_inside_
+         << " px_end_top_=" << px_end_top_ << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+    // get px relative position
+    auto rel_px = price_to_relative(px);
+    if (top_levels_.at(rel_px) == 0) {
+      std::ostringstream os;
+      os << "array_based_order_book::reduce_order."
+         << " Trying to reduce a non-existing top_levels_ price"
+         << " (top_levels_[rel_px] == 0)."
+         << " px_begin_top_=" << px_begin_top_ << " px_inside_=" << px_inside_
+         << " px_end_top_=" << px_end_top_ << " px relative position=" << rel_px
+         << " px=" << px << " qty=" << qty;
+      throw jb::feed_error(os.str());
+    }
+
+    // ... reduce the quantity ...
+    top_levels_.at(rel_px) -= qty;
+    if (top_levels_.at(rel_px) < 0) {
+      // ... this is "Not Good[tm]", somehow we missed an order or
+      // processed a delete twice ...
+      JB_LOG(warning) << "negative quantity in order book";
+      top_levels_.at(rel_px) = 0; // cant't be negative
+    }
+    // ... if it is not the inside we are done
+    if (px != px_inside_) {
+      return false;
+    }
+    // Is inside, ... now check if it was removed
+    if (top_levels_.at(rel_px) == 0) {
+      // gets the new inside (if any)
+      px_inside_ = next_best_price();
+      if (px_inside_ == empty_quote().first) {
+        // last top_levels_ price was removed...
+        // ... get the new inside from the bottom_levels
+        if (!bottom_levels_.empty()) {
+          auto new_inside_it = bottom_levels_.begin();
+          px_inside_ = new_inside_it->first;
+        }
+        // redefine limits
+        auto limits = get_limit_top_prices(px_inside_);
+        px_begin_top_ = std::get<0>(limits);
+        px_end_top_ = std::get<1>(limits);
+        // move tail prices from bottom_levels (px_begin_top_ {, .begin()})
+        move_bottom_to_top();
+      }
+    }
+    return true; // it is an inside change
+  }
+
+  /**
+   * Testing hook.
+   * @returns true is px1 < px2
+   * To test different implementations for buy and sell sides
+   */
+  bool check_better(price4_t const& px1, price4_t const& px2) const {
+    return better_(px1, px2);
+  }
+
+  /** Testing hook.
+   * @returns {px_begin_top_, px_end_top_}
+   */
+  auto get_limits() const {
+    return std::make_tuple(px_begin_top_, px_end_top_);
+  }
+
+  /** Testing hook.
+   * @returns {bottom_levels_ size , top_levels_ size}
+   */
+  auto get_counts() const {
+    return std::make_pair(bottom_levels_.size(), top_levels_count());
+  }
+
+  /** Testing hooks to increase coverage
```

Comment:
ACK.

There are some cases that I can't test if I do not access those function members directly. Examples are throws that are not going to happen becuase the logic of the program does not allow it..... until they happen, right!. So I tested them. The other problem is that without testing them the coverage does not pass. 
What do you want me to do? If I remove them, the coverage will be small.